### PR TITLE
Feature/gal 440 insert new record population

### PR DESCRIPTION
**Pull request by zainm97**
## Description
https://nhsd-jira.digital.nhs.uk/browse/GAL-642
https://nhsd-jira.digital.nhs.uk/browse/GAL-206
https://nhsd-jira.digital.nhs.uk/browse/GAL-440

This PR addresses above three tickets.
Please refer to the diagram in the attachment for ticket 440. This PR covers the top two branches in the workflow.

When a new validated caas feed file is available in the specified s3 bucket, caasFeedAddRecordsLambda is executed and using the AC1, AC2, AC3, AC4 in GAL-440 the record is processed.

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "csv": "^6.3.6",
     "csv-parser": "^3.0.0",
     "csv-stringify": "^6.4.4",
-    "loglevel": "^1.8.1"
+    "loglevel": "^1.8.1",
```
[the remark this answers is not in the document]
Comment (Amar-NHS) [reply package.json:27]: Don't mind this comment as realised this package file is not associated with your Lambda.

File: terraform/modules/iam_galleri_role/iam.tf
[the remark this answers is not in the document]
Comment (Amar-NHS) [reply terraform/modules/iam_galleri_role/iam.tf:1]: As discussed, will only need role policy block and commented out for now which is done.

File: terraform/main.tf [hunk 2]
```diff
@@ -641,6 +649,36 @@ module "user_accounts_lambda_trigger" {
...
+  environment          = var.environment
+  bucket_id            = module.s3_bucket.bucket_id
+  lambda_iam_role      = module.iam_galleri_lambda_role.galleri_lambda_role_arn
+  lambda_function_name = "caasFeedAddRecordsLambda"
+  lambda_timeout       = 100
+  memory_size          = 1024
+  lambda_s3_object_key = "caas_feed_add_records_lambda.zip"
+  environment_vars = {
+    BUCKET_NAME = "galleri-caas-data",
+    ENVIRONMENT = "${var.environment}"
```
[the remark this answers is not in the document]
Comment (zainm97) [reply terraform/main.tf:663]: done

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+} from "@aws-sdk/client-dynamodb";
+import RandExp from 'randexp'
+import { Readable } from "stream";
+import csv from "csv-parser";
+
+const s3 = new S3Client();
+const client = new DynamoDBClient({ region: "eu-west-2", convertEmptyValues: true });
+
+const ENVIRONMENT = process.env.ENVIRONMENT
+const BUCKET = process.env.BUCKET_NAME
```
[the remark this answers is not in the document]
Comment (zainm97) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:20]: removed

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+
+// Lambda Entry Point
+export const handler = async (event) => {
+
+  const bucket = event.Records[0].s3.bucket.name;
+  const key = decodeURIComponent(event.Records[0].s3.object.key.replace(/\+/g, ' '));
+  console.log(`Triggered by object ${key} in bucket ${bucket}`);
+
+  try {
+    const start = Date.now();
```
[the remark this answers is not in the document]
Comment (zainm97) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:35]: removed

**Review by Abdul-NHS — APPROVED** [review]
Really good work!

**Review by Amar-NHS — APPROVED** [review]
looks good. Approved

**Pull request merged**